File: courses/03-skills-competencies/validate.mjs
import fs from 'node:fs';
import path from 'node:path';
import {fileURLToPath} from 'node:url';
const dir=path.dirname(fileURLToPath(import.meta.url));
const doc=JSON.parse(fs.readFileSync(path.join(dir,'references.json'),'utf8'));
const stub=JSON.parse(fs.readFileSync(path.join(dir,'references.stub.json'),'utf8'));
const manifest=JSON.parse(fs.readFileSync(path.join(dir,'assets/manifest.json'),'utf8'));
const failures=[];const ids=new Set();const groups=new Map();
if(JSON.stringify(doc.questions.map(q=>[q.id,q.title]))!==JSON.stringify(stub.questions.map(q=>[q.id,q.title])))failures.push('Question titles differ from prompt');
for(const q of doc.questions)if(!q.answer?.trim())failures.push(q.id+' missing answer');
for(const r of doc.references){
 if(ids.has(r.id))failures.push('Duplicate '+r.id);ids.add(r.id);
 if(r.selected!==false)failures.push('Selected '+r.id);
 if(!r.take||!r.images.length||r.images.length>8)failures.push('Incomplete '+r.id);
 const key=r.app.toLowerCase()+'|'+r.question;const sources=groups.get(key)||new Set();sources.add(r.source);groups.set(key,sources);
 for(let i=0;i<r.images.length;i++)if(!manifest[r.id]?.[i]||!fs.existsSync(path.join(dir,'assets',manifest[r.id][i])))failures.push('Missing image '+r.id+'#'+i);
}
for(const [key,sources]of groups)if(sources.size>1)failures.push('Multiple sources '+key);
const synthesis=fs.readFileSync(path.join(dir,'synthesis.html'),'utf8');
for(const [,id] of synthesis.matchAll(/href="#((?:mobbin|refero):[^"]+)"/g))if(!ids.has(id))failures.push('Bad synthesis ref '+id);
const counters=doc.references.filter(r=>r.counter_example).length;
if(counters<1||counters>3)failures.push('Counter-example count '+counters);
if(failures.length)throw Error(failures.join('\n'));
console.log(JSON.stringify({references:ids.size,products:new Set(doc.references.map(r=>r.app)).size,counters,questions:doc.questions.length,valid:true}));

File: courses/07-share-with-participants/validate.py
import json,pathlib,collections
p=pathlib.Path(__file__).parent
x=json.loads((p/'references.json').read_text()); refs=x['references']; ids=[r['id'] for r in refs]
assert len(ids)==len(set(ids)), 'Duplicate ids'
assert x['module']=='07-share-with-participants'
assert len(x['questions'])==6
assert all(q['answer'] for q in x['questions'])
assert all(r['selected'] is False and r['images'] and len(r['images'])<=8 for r in refs)
assert all(r['question'] in {q['id'] for q in x['questions']} for r in refs)
assert all(r['source'] in ['refero','mobbin'] for r in refs)
groups=collections.defaultdict(set)
for r in refs: groups[(r['app'].lower(),r['question'])].add(r['source'])
assert all(len(s)==1 for s in groups.values())
m=json.loads((p/'assets/manifest.json').read_text())
assert all(len(m[r['id']])==len(r['images']) and all(a and (p/'assets'/a).is_file() for a in m[r['id']]) for r in refs)
assert 1<=sum(r['counter_example'] for r in refs)<=3
print(json.dumps({'references':len(refs),'products':len(set(r['app'] for r in refs)),'per_question':dict(collections.Counter(r['question'] for r in refs)),'counter_examples':sum(r['counter_example'] for r in refs)},indent=2))
